modules/apps/transfer/keeper/forwarding.go

```diff
@@ -0,0 +1,97 @@
+package keeper
+
+import (
+	"errors"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/v8/modules/core/24-host"
+)
+
+// onForwardedPacketErrorAck reverts the receive packet logic that occurs in the middle chain and asynchronously acknowledges the prevPacket
+func (k Keeper) onForwardedPacketErrorAck(ctx sdk.Context, prevPacket channeltypes.Packet, failedPacketData types.FungibleTokenPacketDataV2) error {
+	// the forwarded packet has failed, thus the funds have been refunded to the intermediate address.
+	// we must revert the changes that came from successfully receiving the tokens on our chain
+	// before propagating the error acknowledgement back to original sender chain
+	if err := k.revertForwardedPacket(ctx, prevPacket, failedPacketData); err != nil {
+		return err
+	}
+
+	forwardAck := channeltypes.NewErrorAcknowledgement(errors.New("forwarded packet failed"))
```

Comment:
All of the error acknowledgements written here will not have any error code or codespace associated with them if we just use `errors.New()`, since string error info is stripped in `NewErrorAckwnoledgement` in 04-channel.

We should consider adding `errorsmod.Wrap(someRegisteredErr)`

Reply:
Should we make this a separate issue? I'm trying to keep this PR as a fix and refactor.

Reply:
I think that is a good idea. I would love to see this merged sooner rather than later to keep momentum going (and it would let more people pitch in on those separate issues) :)

Reply:
https://github.com/cosmos/ibc-go/issues/6586